File: compiler/nodes/expressions/call.go
package expressions

import (
	"wipple/codegen"
	"wipple/database"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type CallExpressionNode struct {
	Function database.Node
	Inputs   []database.Node
	Facts    *database.Facts

	isUnit bool
}

func (node *CallExpressionNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseCallExpression(parser *syntax.Parser) (*CallExpressionNode, *syntax.Error) {
	span := parser.Spanned()

	functionValue, err := ParseAtomicExpression(parser)
	if err != nil {
		return nil, err
	}

	many, err := syntax.ParseMany(parser, 1, ParseAtomicExpression, syntax.ParseNothing)
	if err != nil {
		return nil, err
	}

	inputs := make([]database.Node, 0, len(many))
	for _, item := range many {
		inputs = append(inputs, item.Value)
	}

	return &CallExpressionNode{
		Function: functionValue,
		Inputs:   inputs,
		Facts:    database.NewFacts(span()),
	}, nil
}

func (node *CallExpressionNode) Visit(visitor *visit.Visitor) {
	visitExpression(visitor, node)

	if len(node.Inputs) == 1 {
		if unit, ok := node.Inputs[0].(*VariableExpressionNode); ok {
			if definitions, ok := visit.Peek[*visit.ConstantDefinition](visitor, unit.Variable); ok {
				unitDefinition := definitions[0]

				if unitDefinition.Attributes.Unit {
					visitor.Visit(unit)
					visitor.Visit(node.Function)

					visitor.Constraint(typecheck.NewTypeConstraint(unit, typecheck.FunctionType[database.Node]([]database.Node{node.Function}, node)))

					node.isUnit = true

					return
				}
			}
		}
	}

	for _, input := range node.Inputs {
		visitor.Visit(input)
	}

	visitor.Visit(node.Function)

	visitor.Constraint(typecheck.NewTypeConstraint(node.Function, typecheck.FunctionType[database.Node](node.Inputs, node)))
}

func (node *CallExpressionNode) Codegen(c *codegen.Codegen) error {
	span := database.GetSpanFact(node)

	if node.isUnit {
		c.WriteString(span, "await (")
		if err := c.Write(node.Inputs[0]); err != nil {
			return err
		}
		c.WriteString(span, ")(")
		if err := c.Write(node.Function); err != nil {
			return err
		}
		c.WriteString(span, ")")

	} else {
		c.WriteString(span, "await (")
		if err := c.Write(node.Function); err != nil {
			return err
		}
		c.WriteString(span, ")(")

		for _, input := range node.Inputs {
			if err := c.Write(input); err != nil {
				return err
			}

			c.WriteString(span, ", ")
		}

		c.WriteString(span, ")")
	}

	return nil
}

File: compiler/src/nodes/expressions/call.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenResult, ir},
    database::{Fact, Node, NodeRef, Render},
    nodes::{VariableExpressionNode, parse_atomic_expression, visit_expression},
    syntax::{ParseError, Parser},
    typecheck::TypeConstraint,
    visit::{Definition, Visit, Visitor},
};

#[derive(Debug, Clone)]
pub struct ResolvedCall {
    pub is_unit: bool,
}

impl Fact for ResolvedCall {}

impl Render for ResolvedCall {}

#[derive(Debug)]
pub struct CallExpressionNode {
    pub function: NodeRef,
    pub inputs: Vec<NodeRef>,
}

impl Node for CallExpressionNode {}

pub fn parse_call_expression(parser: &mut Parser<'_>) -> Result<CallExpressionNode, ParseError> {
    let function = parse_atomic_expression(parser)?;

    let inputs = parser
        .parse_many(1, parse_atomic_expression, |parser| parser.parse_nothing())?
        .into_iter()
        .map(|(node, _)| node)
        .collect();

    Ok(CallExpressionNode { function, inputs })
}

impl Visit for CallExpressionNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_expression(node, visitor);

        if self.inputs.len() == 1
            && let Some(variable) = self.inputs[0].downcast_ref::<VariableExpressionNode>()
        {
            let unit_constant = visitor
                .peek(&variable.variable, |definition| match definition {
                    Definition::Constant(definition) => Some(definition.clone()),
                    _ => None,
                })
                .find(|definition| definition.attributes.unit);

            if unit_constant.is_some() {
                let input = &self.inputs[0];

                visitor.visit(input);
                visitor.visit(&self.function);

                visitor.constraint(TypeConstraint::new(
                    input.clone(),
                    visitor
                        .db
                        .function_type([self.function.clone()], node.clone()),
                ));

                visitor.edge(&self.function, node, "function");
                visitor.edge(input, node, "input");

                visitor.insert(node, ResolvedCall { is_unit: true });

                return;
            }
        }

        visitor.visit(&self.function);
        visitor.edge(&self.function, node, "function");

        for input in &self.inputs {
            visitor.visit(input);
            visitor.edge(input, node, "input");
        }

        visitor.constraint(TypeConstraint::new(
            self.function.clone(),
            visitor
                .db
                .function_type(self.inputs.iter().cloned(), node.clone()),
        ));

        visitor.insert(node, ResolvedCall { is_unit: false });
    }
}

impl Codegen for CallExpressionNode {
    fn codegen(&self, node: &NodeRef, ctx: &mut CodegenCtx<'_>) -> CodegenResult {
        let ResolvedCall { is_unit } = ctx
            .get(node)
            .ok_or_else(|| anyhow::format_err!("unresolved"))?;

        if is_unit {
            ctx.codegen(&self.inputs[0])?;

            ctx.codegen(&self.function)?;

            ctx.instruction(ir::Instruction::Value {
                node: node.clone(),
                value: ir::Value::Call {
                    function: self.inputs[0].clone(),
                    inputs: vec![self.function.clone()],
                },
            });
        } else {
            ctx.codegen(&self.function)?;

            for input in &self.inputs {
                ctx.codegen(input)?;
            }

            ctx.instruction(ir::Instruction::Value {
                node: node.clone(),
                value: ir::Value::Call {
                    function: self.function.clone(),
                    inputs: self.inputs.clone(),
                },
            });
        }

        Ok(())
    }
}
